### ISSUE TITLE
Subscribing to characteristics using handles

### ISSUE THREAD

**Issue by javl** [noble/noble#826]
I don't like using the issue tracker for questions, but as I'm not getting any responses [on StackOverflow](https://stackoverflow.com/questions/52030336/subscribing-to-characteristics-using-handles-using-noble-node) I'll give it a change:

In order to communicate with a device I'm now using ```peripheral.discoverServices``` and ```peripheral.discoverCharacteristics``` to discover all available characteristics, and I match those to a list of known characteristics UUIDs (as per the pizza example in the Noble repository).

This means I connect to my device, discover and match the characteristics, write and read from the ones I need (including subscribing to a characteristic that needs some time to return a response, preventing the need for polling).

This all works as expected. But as I need to poll 30+ different devices at regular intervals it feels a bit odd to have to go through the discovering part every time and for every device.

In an attempt to skip the discovering and speed up the process I tried using the characteristics' handles (which are the same for all devices). They allow me to communicate straight after connecting to the device using ```peripheral.readHandle(...)``` and ```peripheral.writeHandle(...)```. instead of having to go through discovering first. But it seems this way only allows reading and writing, not subscribing, which I really prefer above polling.

So the question is; is there a way to subscribe to a characteristic through its handle, or to get a handle's characteristic directly so I can subscribe to it, instead of having to go through discovering every time?

**Comment by berningm** [comment]
Actually enabling notify on a characteristic is just another write to the appropriate handle. If you use a tool to read the full GATT profile, you will see that any characteristic  (handle X, e.g. 0x000e) which supports notifications will have one handle (usually X+1, e.g. 0x000f) for the value and another handle (usually X+1+1, e.g. 0x0010) for the notifications. This is the Client Characteristic Configuration Descriptor (CCCD).The +1 is a convention and not necessarily true for all implementations.

Writing 0x0001 to the CCCD will enable notifications; 0x0002 for indications if supported. Writing 0x0000 disables them. Note that according to the standard these are single bit flags, so you could overwrite some other configuration by writing the full value without reading it first. This is very unlikely if you are the only one interacting with this BLE device.

This can also be traced when you look at the code which implements notifications:
https://github.com/noble/noble/blob/c4cd18a7a429bb832f6c4ca793be3faf9af884e9/lib/hci-socket/gatt.js#L617-L656

**Comment by javl** [comment]
Thank you for the reply. 
In the end I just did it like I described: I request all characteristics and match them to a list of known characteristics. I still think it's kind of weird to 'search' for the characteristics every time, but it seems the best way to go about things when using the standard Noble options. Also, hardcoding handles might come back to bite me when one of the devices turns out to be running other firmware or something like that.
